spock-core/src/main/java/org/spockframework/mock/runtime/ProxyBasedMockFactory.java

```diff
@@ -74,9 +90,47 @@ private Object createDynamicProxyMock(Class<?> mockType, List<Class<?>> addition
   }
 
   // inner class to defer class loading
-  private static class CglibMockFactory {
+  private static class ByteBuddyMockFactory {
+
+    private static final ByteBuddyClassCache CACHE = new ByteBuddyClassCache(true);
+
     static Object createMock(Class<?> type, List<Class<?>> additionalInterfaces, @Nullable List<Object> constructorArgs,
         IProxyBasedMockInterceptor interceptor, ClassLoader classLoader, boolean useObjenesis) {
+      Class<?> enhancedType = CACHE.find(classLoader, type, additionalInterfaces);
+
+      if (enhancedType == null) {
+        synchronized (type) {
```

Comment:
Typically, people create a single mock type per class. The alternative where Byte Buddy would need to create multiple mock types for mocking a type plus additional (explicitly specified) interfaces is rather the exception. Given that most mock types derive of a single type, the application is lock-free for these scenarios. Covering the 99% scenario lock free is probably a sane approach. Also, this requires locking only a single monitor such that dead locks are impossible (live locks are however still possible).

Locking on the class loader is more likely to cause problems, though, as class loaders typically lock on themselves during class loading. Creating a mock can cause class loading as using the reflection API (for finding methods to override) causes eager resolution of a type where all types references in methods are loaded. This is typically not a problem but can be one with parallel capable class loaders.